include/xtensor/xmath.hpp

```diff
@@ -2712,7 +2716,11 @@ namespace detail {
         auto inner_mean = nanmean<result_type>(sc, std::move(axes_copy));
 
         // fake keep_dims = 1
-        auto keep_dim_shape = e.shape();
+        // Since the inner_shape might have a reference semantic (e.g. xbuffer_adaptor in bindings)
+        // We need to map it to another type before modifying it.
+        // We pragmatically abuse `get_strides_t`
+        using tmp_shape_t = get_strides_t<typename std::decay_t<E>::shape_type>;
+        tmp_shape_t keep_dim_shape = copy_to<tmp_shape_t>(e.shape().begin(), e.shape().end());
```

Comment:
You can use [xtl::forward_sequence<tm_shape_t>](https://github.com/xtensor-stack/xtl/blob/master/include/xtl/xsequence.hpp#L171) instead of `copy_to` (and thus you remove this function)